source/Ship.cpp

```diff
@@ -2049,27 +2063,22 @@ void Ship::DoGeneration()
 		
 		double coolingEfficiency = CoolingEfficiency();
 		energy += attributes.Get("energy generation") - attributes.Get("energy consumption");
-		fuel += attributes.Get("fuel generation");
+
+		double fuelGeneration = attributes.Get("fuel generation");
+		double fuelGenEnergy = attributes.Get("fuel generation energy") / fuelGeneration;
+		double fuelGenHeat = attributes.Get("fuel generation heat") / fuelGeneration;
+		DoRepair(fuel, fuelGeneration, attributes.Get("fuel capacity"), energy, fuelGenEnergy, fuel, 0., heat, fuelGenHeat);
+		
 		heat += attributes.Get("heat generation");
 		heat -= coolingEfficiency * attributes.Get("cooling");
 		
 		// Convert fuel into energy and heat only when the required amount of fuel is available.
-		if((attributes.Get("fuel consumption") + attributes.Get("regenerative fuel consumption")) <= fuel)
-		{
-			// Also, only convert fuel if it won't create more power than the batteries can hold.
-			double scale = min(1., (attributes.Get("energy capacity") - energy) / attributes.Get("fuel energy"));
-			scale = max(-1., scale);
-			// "regenerative fuel consumption" can run in reverse
-			if (scale < 0)
-				fuel -= scale * attributes.Get("regenerative fuel consumption");
-			else
-				fuel -= scale * (attributes.Get("fuel consumption") + attributes.Get("regenerative fuel consumption"));
-			// the only case where this fails is if no fuel consumption has occurred in the first place
-			if(scale >= 0 || attributes.Get("regenerative fuel consumption"))
-			{
-				energy += scale * attributes.Get("fuel energy");
-				heat += max(0., scale * attributes.Get("fuel heat"));
-			}
+		if(attributes.Get("fuel consumption") <= fuel)
+		{
+			double fuelEnergy = attributes.Get("fuel energy");
+			double fuelConsumption = attributes.Get("fuel consumption") / fuelEnergy;
+			double fuelHeat = attributes.Get("fuel heat") / fuelEnergy;
+			DoRepair(energy, fuelEnergy, attributes.Get("energy capacity") + fuelEnergy, energy, 0., fuel, fuelConsumption, heat, fuelHeat);
 		}
```

Comment:
DoRepair already ensures that the attributes are within allowable bounds, so you don't need the if statement. Also, after some more thinking, it might be better to choose the max between the energy capacity and the fuel energy instead of adding the two. This way we're only overcharging the ship when we need to (i.e. no wasted fuel). This also effectively removes the need for the conversion of energy back into fuel, at least for battery loadouts.
```
DoRepair(energy, fuelEnergy, max(attributes.Get("energy capacity"), fuelEnergy), energy, 0., fuel, fuelConsumption, heat, fuelHeat);
```